File: src/main/java/com/iminling/model/common/ResultModel.java
package com.iminling.model.common;

import com.baomidou.mybatisplus.core.metadata.IPage;
import com.iminling.model.page.Pagination;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.experimental.Accessors;

/**
 * 统一返回实体
 * @author yslao@outlook.com
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@ToString
public class ResultModel<T> {

    /**
     * 状态码
     * @default 200
     */
    private int code;

    /**
     * 返回的数据
     */
    private T data;

    /**
     * 返回的消息
     */
    private String message;

    public ResultModel(MessageCode messageCode){
        this.code = messageCode.getCode();
        this.message = messageCode.getMessage();
    }

    public static ResultModel isOk(){
        return new ResultModel();
    }

    public static <T> ResultModel isOk(T data){
        ResultModel resultModel = new ResultModel();
        resultModel.setData(data);
        return resultModel;
    }

    public static ResultModel isFail(){
        return isFail(MessageCode.RESULT_FAIL);
    }

    public static ResultModel isFail(Integer code){
        return new ResultModel().setCode(code);
    }

    public static ResultModel isFail(Throwable e){
        return isFail().setMessage(e.getMessage());
    }

    public static ResultModel isFail(MessageCode messageCode){
        return new ResultModel()
                .setCode(messageCode.getCode())
                .setMessage(messageCode.getMessage());
    }

    public static ResultModel isFail(String message){
        return new ResultModel()
                .setCode(MessageCode.RESULT_FAIL.getCode())
                .setMessage(message);
    }

    public static ResultModel isFail(String message, int httpStatus){
        return new ResultModel().setCode(httpStatus).setMessage(message);
    }

    public ResultModel setData(T data){
        if (data instanceof IPage){
            IPage page = (IPage) data;
            Pagination pagination = new Pagination<>(page);
            this.data = (T) pagination;
        } else {
            this.data = data;
        }
        return this;
    }
}

File: iminling-core/src/main/kotlin/com/iminling/core/config/value/ResultModel.kt
package com.iminling.core.config.value

import com.iminling.core.constant.MessageCode

/**
 * 统一返回实体
 * @author  yslao@outlook.com
 * @since  2021/11/26
 */
class ResultModel<T> {

    constructor()

    constructor(messageCode: MessageCode): this() {
        this.code = messageCode.code
        this.message = messageCode.message
    }

    /**
     * 状态码
     * @default 200
     */
    var code = 0

    /**
     * 返回的数据
     */
    var data: T? = null

    /**
     * 返回的消息
     */
    var message: String? = null

    companion object {
        fun ok(): ResultModel<Any> {
            return ResultModel(MessageCode.RESULT_OK)
        }

        fun fail(): ResultModel<Any> {
            return ResultModel(MessageCode.RESULT_FAIL)
        }

        fun fail(message: String): ResultModel<Any> {
            var resultModel = ResultModel<Any>(MessageCode.RESULT_FAIL)
            resultModel.message = message
            return resultModel
        }

        fun fail(messageCode: MessageCode): ResultModel<Any> {
            return ResultModel(messageCode)
        }

        fun fail(message: String, code: Int): ResultModel<Any> {
            var resultModel = ResultModel<Any>()
            resultModel.message = message
            resultModel.code = code
            return resultModel
        }
    }

}